### DISCUSSION TITLE
Local Links should include dynamic titles too?

### DISCUSSION THREAD

**Discussion by markadrake** [umbraco/Umbraco-CMS#11881] [Features and ideas]
[MiloW on Umbraco Forums has identified something that I'm surprised I haven't come across before](https://our.umbraco.com/forum/using-umbraco-and-getting-started/108017-dynamic-links-with-dynamic-title-and-even-label#comment-108017). 

**Is there already a solution I don't know about?**

# The Issue

He is working on a multi-lingual site and is using an RTE to link text to a local node in Umbraco. This is referred to [in the code base as a](/umbraco/Umbraco-CMS/blob/5bfab13dc5a268714aad2426a2b68ab5561a6407/src/Umbraco.Core/Templates/HtmlLocalLinkParser.cs) `localLink` and is obvious when you see the syntax:

```
{localLink:umb://document/some-guid-here}
```

The issue MiloW points out is that the title / label for the link is not dynamic as well. The issue is obvious when you are working in a multi-lingual site and the titles of each page are slightly different.

## The simplest user story is:

- **WHEN** I link text in the rich text editor
   - **IF** the link is local to Umbraco
   - **AND** I didn't manually set the title myself
      - **I WANT** the title of the link to be dynamic

## Current behavior

There are property value converters for [text](https://github.com/umbraco/Umbraco-CMS/blob/5bfab13dc5a268714aad2426a2b68ab5561a6407/src/Umbraco.Core/PropertyEditors/TextStringValueConverter.cs#L22), [grid](https://github.com/umbraco/Umbraco-CMS/blob/5bfab13dc5a268714aad2426a2b68ab5561a6407/src/Umbraco.Infrastructure/PropertyEditors/GridPropertyEditor.cs#L40), [markdown](https://github.com/umbraco/Umbraco-CMS/blob/5bfab13dc5a268714aad2426a2b68ab5561a6407/src/Umbraco.Infrastructure/PropertyEditors/ValueConverters/MarkdownEditorValueConverter.cs#L15), [rich text](https://github.com/umbraco/Umbraco-CMS/blob/5bfab13dc5a268714aad2426a2b68ab5561a6407/src/Umbraco.Infrastructure/PropertyEditors/RichTextPropertyEditor.cs#L38), and [rte macros](https://github.com/umbraco/Umbraco-CMS/blob/5bfab13dc5a268714aad2426a2b68ab5561a6407/src/Umbraco.Infrastructure/PropertyEditors/ValueConverters/RteMacroRenderingValueConverter.cs#L28) that support parsing these values to their HREF/URL equivalents.

There's also the backoffice UI to consider. When you leave the values blank for link, anchor, link title, etc and simply select a content node... These values are then pre-filled.

![image](https://user-images.githubusercontent.com/713355/149990505-9d131268-e36f-4a29-bea0-2838c6cc18eb.png)
![image](https://user-images.githubusercontent.com/713355/149990561-c1211cba-27fa-4ac8-8a2e-aedd6b5f0316.png)

The example above results in the following markup:

```
<a href="/{localLink:umb://document/26ec07a2a35c43218b0f3e79ae7517a3}" 
     title="NSVSAC.org">
     NSVSAC.org
</a>
```

## Proposed Behavior

To jump start this conversation, if we were to implement other dynamic values perhaps we could do something like this:

```
<a href="/{localLink:umb://document/26ec07a2a35c43218b0f3e79ae7517a3}" 
     title="{localTitle:umb://document/26ec07a2a35c43218b0f3e79ae7517a3}">
     NSVSAC.org
</a>
```

If `localTitle` is discovered we would replace it with the current content node's title as it stands for the current language/culture. There a host of other UI blunders ahead. Say if someone decides to edit the text for the link in the RTE, do we respect that or throw it away? Fun!

**Comment by skttl** [comment]
Could make it 
```html
<a href="/{localLink:umb://document/26ec07a2a35c43218b0f3e79ae7517a3}" 
     title="{localName:umb://document/26ec07a2a35c43218b0f3e79ae7517a3}">
     {localName:umb://document/26ec07a2a35c43218b0f3e79ae7517a3}
</a>
```

This could also open for something like `{localProperty:umb://document/26ec07a2a35c43218b0f3e79ae7517a3/propertyAlias}` or `{localDictionary:Key.Of.The.Dictionary.Item}`

**Reply by markadrake** [reply]
Yeah I actually wrote _some very terrible code_ to do dynamic string replacements. This does walk really close to the line of using properties from the object, properties on the document type, and even dictionary values. I'd really like to insert any of these into a text or rich text editor. Ditto!
